### ISSUE TITLE
Issues trying to fit a basic model in Bean Machine

### ISSUE THREAD

**Issue by dougj892** [facebookresearch/beanmachine#1305]
Hi - I just came across Bean Machine on Twitter. This looks really cool! Thanks for all the hard work that went into this.

I was curious about Bean Machine and so took stab at implementing a simple IRT model in Bean Machine. (See blog post here: I wrote a quick blog post about my experience here: https://www.dougjohnson.in/post/stan-vs-pymc3-vs-bean-machine/) Unfortunately, I don't think I got the model to work (the posterior distribution is quite a bit different from what I get when using Stan or PyMC3) and I found some of the documentation a bit confusing. I am definitely not a great programmer so my confusion could entirely be due to my lack of programming experience but I thought I'd share a bit about what I found confusing below in case it is helpful. 

1. When creating families of random variables, I was a bit confused when to use ".expand()" (as you guys do in the radon tutorial) and when to pass an index to the variable function definition (as you guy do in several of the other tutorials). More generally, I found the latter syntax pretty confusing. 
2. I wasn't clear what the role of the function annotation "-> RVIdentifier". This was included in some of the examples but not the others.

Finally, it would also be really cool if you allowed users to "step through" the prior predictive by first sampling from the priors and then from each random or deterministic variable in turn. This would be really useful for basic debugging and could be a big advantage of Bean Machine over Stan / PyMC3. (And apologies if this is possible, e.g. by creating a new World() instance, and I just missed it.)

### Issue Description
A clear and concise description of the issue. If it's a feature request, please add [Feature Request] to the title.

### Steps to Reproduce
Please provide steps to reproduce the issue attaching any error messages and stack traces.

### Expected Behavior
What did you expect to happen?

### System Info
Please provide information about your setup
- PyTorch Version (run `print(torch.__version__)`
- Python version

**Comment by zaxtax** [comment]
Hi Doug! Thanks for giving Bean Machine a spin.

So the difference between the two uses is that when we use `.expand`  and passing an index is that in the former case bean machine sees it as a single random variable, while in the latter there is a unique random variable for every possible index value. As sampling in batch is significantly faster, I encourage you to use `.expand` where you can and use index variables in places where it's less clear how to vectorize the code.

The type annotation is just to make it clear the type of function with the annotation. It is purely optional.

Finally, for a fair comparison you actually want to use `bm.GlobalNoUTurnSampler`. Can you tell me if the following gives you better results:

````
samples = bm.GlobalNoUTurnSampler().infer(
    queries=[theta()],
    observations={y(): torch.tensor(df["response"].values, dtype=torch.float32)},
     num_samples=1000,
     num_chains=4,
     num_adaptive_samples=1000
)
````

**Comment by dougj892** [comment]
Hi Zaxtax -- Thanks a lot for the clarifications on the use of .expand vs passing an index and the type annotation! That is really helpful. Wrt the sampler used, I couldn't seem to get NUTS to work on my model. If I swap out the NUTS sampler for the MH one the progress bar shows "Samples collected: 0%" and doesn't seem to update. The same thing happens if I use basic HMC.

**Comment by zaxtax** [comment]
Did you try explicitly setting the dtype of the observation tensor to be a float? While it's a bit unintuitive since conceptually we think of `dist.Bernoulli` as having support `0` and `1`, it really producing `0.0` and `1.0` when it is sampled.

**Comment by ndmlny-qs** [comment]
@dougj892 I can confirm that @zaxtax's solution fixes the issue. See below for a sampling of `student ability (θ)` posteriors from all three PPLs. I looked at the posteriors for the random variables `question difficulty (b)`, `question discriminator (a)`, and `guess (c)` and all were in close agreement.

![ppl-posteriors](https://user-images.githubusercontent.com/60523020/148617839-1f6c1754-9edb-4bba-9755-c6c58bbe1238.gif)

The important take-away here is that the observations dictionary should have values that are torch tensors, and the items in the tensors should not be integers.

**Comment by dougj892** [comment]
@zaxtax and @ndmlny-qs - Thanks so much for the guidance! Changing the dtype works for me as well. Apologies if this is a dumb question, but I guess the general rule then is that the dtype of the tensor elements passed in the observations dictionary must match the dtype that the relevant torch distribution outputs. Is that correct?

Also, I have made the Colab notebooks I used to test the models publicly accessible. Links below. (And sorry for not making them public earlier.)

Bean Machine colab notebook: https://colab.research.google.com/drive/1clOIXN5pgKtkzRIt2WvQ6s9q2FSbX6-g?usp=sharing

PyMC3 Shareable link: https://colab.research.google.com/drive/15SvGVsplzp_mGub84J8ttS2k8HF0Q94a?usp=sharing
